src/gc/gc.d

```diff
@@ -551,7 +626,25 @@ struct GC
     }
 
     /**
+     * Request that the GC reallocate a block of memory, attempting to adjust
+     * the size in place if possible. If size is 0, the memory will be freed.
+     *
+     * If p was not allocated by the GC, points inside a block, or is null, no
+     * action will be taken.
+     *
+     * Params:
+     *  p = A pointer to the root of a valid memory block or to null.
+     *  size = The desired allocation size in bytes.
+     *  bits = A bitmask of the attributes to set on this block.
+     *  alloc_size = The actuall size allocated in bytes.
+     *  ti = TypeInfo to describe the memory.
+     *
+     * Returns:
+     *  A reference to the allocated memory on success or null if size is
+     *  zero.  On failure, the original value of p is returned.
```

Comment:
> This is from the user facing documentation. Does that mean that should be changed as well?

Not sure. Implementation and documenttion don't match, but I don't think realloc is well defined to begin with.

> On failure, the original value of p is returned.

If "failure" means an "invalid" pointer is passed, this is not a good return value, as it is the same if the memory has been successfully resized in place. "invalid" pointers are not detected at the moment.

> Throws: OutOfMemoryError on allocation failure.

This is the only current failure.

In addition I don't like that reallocating with size 0 frees the memory immediately, but leaves it for garbage collection for other sizes. This will be surprising in generic code.